tests/rpc/gossamer_utils.go

```
@@ -0,0 +1,165 @@
+	//	"--genesis", filepath.Join(currentDir, "../..", "node/gssmr/genesis.json"),
+	//	"--force",
+	//)
+	//
+	////add step for init
+	//t.Log("Going to init gossamer", "cmd", cmd)
+	//err = cmd.Start()
+	//if err != nil {
+	//	t.Error("Could not init gossamer", "err", err)
+	//	return nil, err
+	//}
+
+	//TODO: could we enable genesis file to be configured via args without init?
+	cmd := exec.Command(gossamerCMD, "--port", "700"+strconv.Itoa(nodeNumb),
+		"--key", keyList[nodeNumb],
+		"--datadir", dataDir+strconv.Itoa(nodeNumb),
+		"--rpchost", GOSSAMER_NODE_HOST,
+		"--rpcport", "854"+strconv.Itoa(nodeNumb),
+		"--rpcmods", "system,author,chain",
+		"--key", keyList[nodeNumb],
+		"--config", currentDir+"/config.toml",
+		"--roles", "4",
+		"--rpc",
+	)
+
+	t.Log("Going to execute gossamer", "cmd", cmd)
+	err = cmd.Start()
+	if err != nil {
+		t.Error("Could not execute gossamer cmd", "err", err)
+		return nil, err
+	}
+
+	t.Log("wait few secs for node to come up", "cmd.Process.Pid", cmd.Process.Pid)
+	var started bool
+
+	for i := 0; i < 10; i++ {
+		time.Sleep(3 * time.Second)
+		if err := CheckFunc(t, "http://"+GOSSAMER_NODE_HOST+":854"+strconv.Itoa(nodeNumb)); err == nil {
+			started = true
+			break
+		} else {
+			t.Log("Waiting for Gossamer to start", "err", err)
+		}
+	}
+	if started {
+		t.Log("Gossamer started :D", "cmd.Process.Pid", cmd.Process.Pid)
+	} else {
+		t.Fatal("Gossamer node never managed to start!")
+
+	}
+
+	return cmd, nil
+}
+
+// CheckFunc check if gossamer node is already started
+func CheckFunc(t *testing.T, gossamerHost string) error {
+	method := "system_health"
+
+	respBody := PostRPC(t, method, gossamerHost)
+
+	target := DecodeRPC(t, respBody, method)
+
+	if !target.(*modules.SystemHealthResponse).Health.ShouldHavePeers {
+		return fmt.Errorf("no peers")
+	}
+
+	//if we get here, we assume it worked :D
+
+	return nil
+}
+
+// KillGossamer kills a instance of gossamer
+func KillGossamer(t *testing.T, cmd *exec.Cmd) error {
```

since `t` is passed in here, do you need to return error or can you just do `t.Fatal()`?

---

hmmm, I want to return the err, if we fatal here and there are more nodes to stop they will remain online.

---

It does seem a bit odd that we are not using `t.Fatal()` but rather returning an error. Will prevent all tests from continuing? If so, it would be preferred to find another way to stop the nodes on error or use `t.Fatal()` and let the tests continue and then show the logs once the test has failed.

---

maybe t.Cleanup() could be of use here, it's similar defer but for an entire set of test functions.. it's a go 1.14 feature though

---

Lets avoid making it more complicated on this PR. If you feel like this is a really strong point to make, raise a ticket and we can tackle it later. For now it will stay like this.